jams/sonify.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python
+# CREATED:2015-12-12 18:20:37 by Brian McFee <brian.mcfee@nyu.edu>
+r'''
+Sonification
+============
+
+.. autosummary::
+    :toctree: generated/
+
+    sonify
+'''
+
+import six
+import numpy as np
+import mir_eval.sonify
+from mir_eval.util import filter_kwargs
+from .eval import coerce_annotation
+from .exceptions import NamespaceError
+
+__all__ = ['sonify']
+
+
+
+def clicks(annotation, sr=22050, length=None, **kwargs):
+    '''Sonify clicks timings
+
+    '''
+
+    interval, _ = annotation.data.to_interval_values()
+
+    return filter_kwargs(mir_eval.sonify.clicks, interval[:, 0],
+                         fs=sr, length=length, **kwargs)
+
+
+def chord(annotation, sr=22050, length=None, **kwargs):
+    '''Sonify chords'''
+
+    intervals, chords = annotation.data.to_interval_values()
+
+    return filter_kwargs(mir_eval.sonify.chords,
+                         chords, intervals,
+                         fs=sr, length=length,
+                         **kwargs)
+
+
+def pitch_hz(annotation, sr=22050, length=None, **kwargs):
+    '''Sonify pitches in Hz'''
+
+    intervals, pitches = annotation.data.to_interval_values()
+
+    # Collapse down to a unique set of frequency values
+    freqs = np.unique(pitches)
+
+    # Build the piano roll
+    pitch_index = {p: i for i, p in enumerate(freqs)}
+    gram = np.zeros((len(freqs), len(pitches)))
+    for t, n in enumerate(pitches):
+        gram[pitch_index[n], t] = 1.0
```

Comment:
> Have I mentioned that I hate this negative frequency convention?

Life sucks.

Also, try sonifying this `pitch_midi` annotation (which gets coerced to a `pitch_hz` annotation by the current code), something's going wrong somewhere (I get really low-frequency sounds and a whole bunch of clicks):

```json
{
  "sandbox": {}, 
  "annotations": [
    {
      "namespace": "pitch_midi", 
      "sandbox": {}, 
      "time": 0, 
      "duration": null, 
      "annotation_metadata": {
        "annotation_tools": "", 
        "curator": {
          "name": "", 
          "email": ""
        }, 
        "annotator": {}, 
        "version": "", 
        "corpus": "", 
        "annotation_rules": "", 
        "validation": "", 
        "data_source": ""
      }, 
      "data": {
        "duration": [
          0.28985500000000003, 
          0.262681, 
          0.108695, 
          0.16304300000000002, 
          0.28985500000000003, 
          0.271739, 
          0.28985500000000003, 
          0.28985500000000003, 
          0.182975
        ], 
        "confidence": [
          1.0, 
          1.0, 
          1.0, 
          1.0, 
          1.0, 
          1.0, 
          1.0, 
          1.0, 
          1.0
        ], 
        "value": [
          63.0, 
          62.0, 
          62.0, 
          60.0, 
          57.0, 
          56.0, 
          60.0, 
          63.0, 
          62.0
        ], 
        "time": [
          0.20833300000000002, 
          0.498188, 
          0.760869, 
          0.869564, 
          1.0326060000000001, 
          1.947461, 
          2.2282580000000003, 
          2.5271700000000004, 
          2.817025
        ]
      }
    }
  ], 
  "file_metadata": {
    "jams_version": "0.2.1", 
    "title": "TRACBNP128F427824_62450a93473b0b0d417420081945652b", 
    "identifiers": {}, 
    "release": "", 
    "duration": 3, 
    "artist": ""
  }
}
```